redis.go

```diff
@@ -126,32 +133,37 @@ func blocking(
 		if done {
 			return
 		}
-		// there is no cond.WaitTimeout(), so hence the the goroutine to wait
-		// for a timeout
-		var (
-			wg     sync.WaitGroup
-			wakeup = make(chan struct{}, 1)
-		)
-		wg.Add(1)
-		go func() {
-			m.signal.Wait()
-			wakeup <- struct{}{}
-			wg.Done()
-		}()
-		select {
-		case <-wakeup:
-		case <-dlc:
-			onTimeout(c)
-			m.signal.Broadcast() // to kill the wakeup go routine
-			wg.Wait()
+
+		if m.Ctx.Err() != nil {
 			return
-		case <-m.Ctx.Done():
-			m.signal.Broadcast() // to kill the wakeup go routine
-			wg.Wait()
+		}
+		if timedOut() {
+			onTimeout(c)
 			return
 		}
-		wg.Wait()
+
+		m.signal.Wait()
+	}
+}
+
+func setCondTimer(sig *sync.Cond, timeout time.Duration, cleanup chan struct{}) func() bool {
```

Comment:
Just had an idea how to make this a bit simpler/more robust (maybe?): Instead of re-doing the timeout check after receiving the signal, share a bool variable that's protected by the mutex. I.e.,

```
func setCondTimer(sig *sync.Cond, timedOut *bool, timeout time.Duration, cleanup chan struct{}) {
   ...
}
```

and use it as

```
bool timedOut // protected by cond.Locker
setCondTimer(cond, &timedOut, timeout, ...)
...
if timedOut {
  onTimeout(c)
```

The advantage being we no longer have any subtle dependency on how `time.Since(start)` behaves in different threads. Probably is a nicer way to package it though.

Reply:
Codes is nicer, but I don't see how timedOut is protected. It's set in setCondTimer when the timeout triggers, which is not protected.

Reply:
Good point, would have to do:

```
	dl := time.NewTimer(timeout)
	defer dl.Stop()
	select {
	case <-dl.C:
		cond.L.Lock()
		*timedOut = true
		sig.Broadcast() // main loop might miss this signal
		cond.L.Unlock()
	case <-cleanup:
	}
```